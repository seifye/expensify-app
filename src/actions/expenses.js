// using universally unique identifier (uuid) library to generate id's for our addExpense action generator
import { v4 as uuidv4 } from "uuid";



// Actions generators
// They are functions that return action objects (applying ES6 object destructuring)

// ADD_EXPENSE
export const addExpense = (
    {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
    } = {}
    ) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuidv4(),
        description,
        note,
        amount,
        createdAt
    }
})

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id

})

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

