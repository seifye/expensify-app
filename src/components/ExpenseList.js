// import React for using JSX
import React from 'react'

// import the connect function to connect the components to the redux store
import { connect } from 'react-redux'

// import other js files => use
import ExpenseListItem from './ExpenseListItem'

// import selectors
import getVisibleExpenses from '../selectors/expenses'


// Regular component unconnected to the redux store
// we export it just in the purpose of testing
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p> No expenses </p>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem {...expense} key={expense.id} />
                })
            )
        }
    </div>
)
// function that returns a props object at which the values of the keys can access the redux store
// Note: as the store changes, this function is automatically going to run getting the new values in the component
const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }    
}

/* using the connect method to connect the regular component with a predefined function to get a new 
component which is now connected to the redux store. We export the connected version of the component */
export default connect(mapStateToProps)(ExpenseList)

