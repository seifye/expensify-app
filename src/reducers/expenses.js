// Reducers
// This function is called a reducer that determines what to do based off of the state and the action 
// Rule 1: Reducers are pure functions which means that the output is only determined by the input.
// Rule 2: In Reducer functions, we never want to directly change the inputs instead, returning an object that represents the new state

/* The expenses state is managed by the expenses reducer while the filters state is managed by the 
Filters Reducer */

// Expenses Reducer 
const expensesReducerDefaultState = []
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense]
        case 'REMOVE_EXPENSE': 
            return state.filter(({ id }) => id !== action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense
                }
            })
        default:
            return state;
    }
}

export default expensesReducer