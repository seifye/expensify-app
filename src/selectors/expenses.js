// import the moment.js API as a date library which is much easier than the Date object in JS
import moment from 'moment'


// Selectors are going to refer to things that actually query the data from the redux store

// getVisibleExpenses Function returns the expenses array after filtering and sorting the data 
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true             
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        // returns 1 (a comes first) or -1 (b comes first)
        if (sortBy === 'date') { 
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })
}

export default getVisibleExpenses