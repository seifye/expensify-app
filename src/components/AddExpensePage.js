// import React for using JSX
import React from 'react'

// import the connect function to connect the components to the redux store
import { connect } from 'react-redux'

// import other js files => use
import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'


// Exporting the unconnected version of the component for testing purposes
export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense))       // This is hard to test because addExpense(expense) is an imported function from outside
        this.props.addExpense(expense)                 // This is much easier to test (using a mock function)
        // Navigate to the dashboard page (homepage)
        this.props.history.push('/')
    }
    render() {
        return (
            <div>
                <h1> Add Expense </h1>
                <ExpenseForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}



// It is a way to return your dispatch functions allowing you to abstract them away from the component
const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
})


/* using the connect method to connect the regular component with the state which is now connected to the redux store
and then we export the connected version of the component. Here we did't pass first argument to connect because
we don't need mapStateToprops, we only need to mapDispatchToProps and to access the built in dispatch prop  */
export default connect(undefined, mapDispatchToProps)(AddExpensePage)