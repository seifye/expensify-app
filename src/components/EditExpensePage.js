// import React for using JSX
import React from 'react'

// import the connect function to connect the components to the redux store
import { connect } from 'react-redux'

import { Link } from 'react-router-dom'

// import other js files => use
import ExpenseForm from './ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses'



/* React router component comes with built in props. The React component which got rendered by a Route 
Component can only access these props */
export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        // this.props.dispatch(editExpense(this.props.expense.id, expense))   // This is hard to test because editExpense() is an imported function from outside
        this.props.editExpense(this.props.expense.id, expense)  // This is much easier to test (using a mock function)
        // Navigate to the home page
        this.props.history.push('/') 
    }
    onRemove = () => {
        // this.props.dispatch(removeExpense({ id: this.props.expense.id }))      // This is hard to test because removeExpense() is an imported function from outside
        this.props.removeExpense({ id: this.props.expense.id })         // This is much easier to test (using a mock function)
        // Navigate to the home page
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                This is from my edit expense page.
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit= {this.onSubmit}
                />
                <button onClick = {this.onRemove} > Remove </button>
            </div>
        )
    }
}



// function that returns a props object at which the values of the keys can access the redux store
// Note: as the store changes, this function is automatically going to run getting the new values in the component
const mapStateToProps = (state, props) => ({
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

// It is a way to return your dispatch functions allowing you to abstract them away from the component
const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
})


/* using the connect method to connect the regular component with a predefined function to get a new 
component which is now connected to the redux store. We export the connected version of the component */
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)