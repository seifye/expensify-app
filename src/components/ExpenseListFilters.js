// import React for using JSX
import React from 'react'

// import the connect function to connect the components to the redux store
import { connect } from 'react-redux'

// import the date picker component which is developed by airbnb
import { DateRangePicker } from 'react-dates'

// import other js files => use
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'



// Regular component unconnected to the redux store. Export the unconnected component for testing
export class ExpenseListFilter extends React.Component {
    state = {
        calendarFocused: null
    }
    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate()
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount()
        }
    }

    render() {
        return (
            <div>
                <input 
                    type='text' 
                    value={this.props.filters.text} 
                    onChange = {this.onTextChange}
                />
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onSortChange}
                >
                    <option value='date'> Date </option>
                    <option value='amount'> Amount </option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates = {true}
                    numberOfMonths= {1}
                    isOutsideRange= {() => false}
                />
            </div>
        )
    }
}



// function that returns a props object at which the values of the keys can access the redux store
// Note: as the store changes, this function is automatically going to run getting the new values in the component
const mapStateToProps = (state) => ({
        filters: state.filters
})

// It is a way to return your dispatch functions allowing you to abstract them away from the component
const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})


/* using the connect method to connect the regular component with a predefined function to get a new 
component which is now connected to the redux store. We export the connected version of the component */
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter)