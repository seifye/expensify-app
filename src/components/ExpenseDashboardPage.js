// import React for using JSX
import React from 'react'

// import other js files => use
import ExpenseList from './ExpenseList'
import ExpenseListFilter from './ExpenseListFilters'


const ExpenseDashboardPage = () => (
    <div>
        <ExpenseListFilter />
        <ExpenseList />
    </div>
)


export default ExpenseDashboardPage