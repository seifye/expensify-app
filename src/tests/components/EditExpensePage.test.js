// import React for using JSX
import React from 'react'

/* There are two main ways to test react components (Shallow rendering and full DOM rendering). 
Shallow rendering renders only a given component while full DOM rendering renders child components as
well */

// Importing the shallow renderer from enzyme library
import { shallow } from 'enzyme'

// import other js files => use
import { EditExpensePage } from '../../components/EditExpensePage'
import expenses from '../fixtures/expenses'


let editExpenseSpy, removeExpenseSpy, historySpy,  wrapper

// this is a function will run before each test case
beforeEach(() => {
    // creating two spy functions to fake the props functions on the addExpense component
    editExpenseSpy = jest.fn()
    removeExpenseSpy = jest.fn()
    historySpy = {
        push: jest.fn()
    }
    wrapper = shallow(
        <EditExpensePage 
            editExpense={editExpenseSpy} 
            removeExpense = {removeExpenseSpy}
            history={historySpy} 
            expense={expenses[2]}
        />
    )
})



// Testing React Component

// Test case to render the EditExpensePage correctly
test('should render the EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot()
})
  

// Test case to check the onSubmit handler
test('should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2])
    expect(historySpy.push).toHaveBeenLastCalledWith('/')
    expect(editExpenseSpy).toHaveBeenLastCalledWith(expenses[2].id, expenses[2])
})


// Test case to check the onRemove handler
test('should handle onRemove', () => {
    wrapper.find('button').simulate('click')
    expect(historySpy.push).toHaveBeenLastCalledWith('/')
    expect(removeExpenseSpy).toHaveBeenLastCalledWith({
        id: expenses[2].id
    })
})
