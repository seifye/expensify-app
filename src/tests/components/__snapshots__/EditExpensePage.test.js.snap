// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the EditExpensePage correctly 1`] = `
<div>
  This is from my edit expense page.
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit card",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
     Remove 
  </button>
</div>
`;
